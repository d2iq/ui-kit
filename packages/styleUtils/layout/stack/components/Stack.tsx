import * as React from "react";
import { SpaceSize } from "../../../../shared/styles/styleUtils/modifiers/modifierUtils";
import { stack } from "../style";
import { cx } from "@emotion/css";
import { listReset } from "../../../../shared/styles/styleUtils";
import { flex } from "../../../../shared/styles/styleUtils/layout/flexbox";

interface StackProps {
  className?: string;
  /**
   * human-readable selector used for writing tests
   */
  ["data-cy"]?: string;
  /**
   * The size of the space to apply between items. It can set 1 spacing size for all breakpoints, or it can be used to set different spacing values at different viewport width breakpoints.
   */
  spacingSize?: SpaceSize;
  /**
   * Which HTML tag to render the component with
   */
  tag?: keyof React.ReactHTML;
}

const Stack: React.StatelessComponent<StackProps> = ({
  children,
  className,
  "data-cy": dataCy,
  spacingSize,
  tag: Tag = "div"
}) => (
  <Tag
    className={cx(
      { [listReset]: Tag === "ul" || Tag === "ol" },
      stack(spacingSize),
      flex({ direction: "column" }),
      className
    )}
    data-cy={dataCy}
  >
    {children}
  </Tag>
);

Stack.defaultProps = {
  spacingSize: "m",
  tag: "div",
  ["data-cy"]: "stack"
};

export default Stack;
