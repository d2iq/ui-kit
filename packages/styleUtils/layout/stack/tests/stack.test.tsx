import React from "react";
import { render } from "enzyme";
import toJSON from "enzyme-to-json";
import Stack from "../components/Stack";

describe("Stack", () => {
  it("renders default", () => {
    expect(
      toJSON(
        render(
          <Stack>
            <div>Lorem Ipsum is simply dummy text</div>
            <div>of the printing and typesetting industry.</div>
            <div>Lorem Ipsum has been the industry's</div>
          </Stack>
        )
      )
    ).toMatchSnapshot();
  });
  it("renders with different spacingSize", () => {
    expect(
      toJSON(
        render(
          <Stack spacingSize="l">
            <div>Lorem Ipsum is simply dummy text</div>
            <div>of the printing and typesetting industry.</div>
            <div>Lorem Ipsum has been the industry's</div>
          </Stack>
        )
      )
    ).toMatchSnapshot();
  });
  it("renders with responsive spacingSize", () => {
    expect(
      toJSON(
        render(
          <Stack spacingSize={{ default: "s", medium: "m" }}>
            <div>Lorem Ipsum is simply dummy text</div>
            <div>of the printing and typesetting industry.</div>
            <div>Lorem Ipsum has been the industry's</div>
          </Stack>
        )
      )
    ).toMatchSnapshot();
  });
  it("renders with a custom tag", () => {
    expect(
      toJSON(
        render(
          <Stack tag="ul">
            <li>Lorem Ipsum is simply dummy text</li>
            <li>of the printing and typesetting industry.</li>
            <li>Lorem Ipsum has been the industry's</li>
          </Stack>
        )
      )
    ).toMatchSnapshot();
  });
});
