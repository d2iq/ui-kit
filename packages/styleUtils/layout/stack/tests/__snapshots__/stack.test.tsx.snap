// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack renders default 1`] = `
<div
  class="css-15twc3j"
  data-cy="stack"
>
  <div>
    Lorem Ipsum is simply dummy text
  </div>
  <div>
    of the printing and typesetting industry.
  </div>
  <div>
    Lorem Ipsum has been the industry's
  </div>
</div>
`;

exports[`Stack renders with a custom tag 1`] = `
<ul
  class="css-15wkwvi"
  data-cy="stack"
>
  <li>
    Lorem Ipsum is simply dummy text
  </li>
  <li>
    of the printing and typesetting industry.
  </li>
  <li>
    Lorem Ipsum has been the industry's
  </li>
</ul>
`;

exports[`Stack renders with different spacingSize 1`] = `
<div
  class="css-p1p1p3"
  data-cy="stack"
>
  <div>
    Lorem Ipsum is simply dummy text
  </div>
  <div>
    of the printing and typesetting industry.
  </div>
  <div>
    Lorem Ipsum has been the industry's
  </div>
</div>
`;

exports[`Stack renders with responsive spacingSize 1`] = `
<div
  class="css-i8pzeb"
  data-cy="stack"
>
  <div>
    Lorem Ipsum is simply dummy text
  </div>
  <div>
    of the printing and typesetting industry.
  </div>
  <div>
    Lorem Ipsum has been the industry's
  </div>
</div>
`;
